functorch/csrc/BatchRulesConv.cpp

```diff
@@ -0,0 +1,83 @@
+#include <functorch/csrc/BatchRulesHelper.h>
+
+namespace at { namespace functorch {
+
+// batching rules translated from jax: https://github.com/google/jax/blob/master/jax/_src/lax/lax.py#L3143
+Tensor reshape_axis_into(int64_t src, int64_t dst, const Tensor &x) {
+  auto new_shape = x.sizes().vec();
+  new_shape.erase(new_shape.begin() + src);
+  new_shape[dst] *= x.sizes()[src];
+  return at::reshape(x.movedim(dst, src), new_shape);
+}
+
+Tensor reshape_axis_outof(int64_t src, int64_t size1, const Tensor &x) {
+  auto shape = x.sizes().vec();
+  TORCH_INTERNAL_ASSERT(shape[src] % size1 == 0);
+  int64_t size2 = shape[src] / size1;
+  shape[src] = size1;
+  shape.insert(shape.begin() + src + 1, size2);
+  return at::reshape(x, shape);
+}
+
+// Does not support batch_group_count (needed for convolution backwards)
+std::tuple<Tensor,optional<int64_t>>
+conv2d_batching_rule(const Tensor& lhs, optional<int64_t> lhs_bdim, const Tensor& rhs, optional<int64_t> rhs_bdim, const optional<Tensor>& bias, optional<int64_t> bias_bdim, IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
+  std::vector<int64_t> lhs_spec = {0,1,2,3};
+  std::vector<int64_t> rhs_spec = {0,1,2,3};
+  std::vector<int64_t> out_spec = {0,1,2,3};
```

Comment:
I feel like I should be able to get rid of these entirely... but I wasn't totally sure (so I didn't).

Reply:
Yeah, you definitely can just hardcode these values in because people only really use NCHW convolutions. I don't understand how MemoryFormat in PyTorch works though and how this would fit into that

Reply:
The reason I left this here is because I'm not sure if we need to do something different for say, channels last convolutions?

Reply:
I'm not sure either. The best way to find out is probably to write a test. Our ChannelsLast tensors have dimensions in NCHW order, but the strides have NHWC layout. My hypothesis is that channels last convolution might turn into a channels first convolution when it goes through this batching rule which would be a performance problem to solve in the future

Reply:
Decided to switch back to vectors, just since it'll probably be easier to extend to general conv.

Reply:
nit: std::vector<int64_t> -> VmapDimVector / DimVector can avoid dynamic allocations